### DISCUSSION TITLE
Multiple deployment management

### DISCUSSION THREAD

**Discussion by phyzical** [argoproj/argo-rollouts#2912] [General]
Hey,

So we have a situation where we have a crd defining a rollout which runs some a backend server for a rails application with HA so X replica, 
* this runs a prePromotionAnalysisTemplate to perform some checks before promoting traffic
* then runs a postPtomotionAnalysisTemplate while traffic is routed.
Which if any fail (or the pod fails to boot) it will rollback to the last working version

and then we have another (currently deployment) that uses the same docker image to run a background worker with a single replica which talks to the same database as the above, reason we dont want X replica for this is we just wont require the additional workers

The Question we have is there a way to have this single replica deployment keep in sync with the image being used by the rollout?

I could not see any way to pull this off via Deployment crd logic, so the closest i could think would be to make the worker into a rollout also and trigger the same set of analysisTemplates.

This would probably work for the most part, but 
* this would incur additional costs rerunning the same checks (which will happen anyway)
* means the checks  run by the worker may run before the main pod is ready leading to false positive rollbacks
* runs the risk that if one incurs a rollback (maybe due to a flaky check) the other may not fail and thus the versioning will be out of sync again

Basically the issue we are trying to avoid surrounds that both use the same codebase specifically  for the sake of its database operations

the obvious solution is just add this as another container in the current rollout for the workers context, but we think this would be overkill in terms of its workload

if this is not a thing yet, would it be a matter of making a new feature request to do something like being able to ask one rollout to use another rollouts analysis results like 
```
prePromotionAnalysisTemplate
   - rolloutRef: ROLLOUT_NAME
   - templateRef: TEMPLATE_NAME
```
though i'm sure its not as simple and just delegating the responsibility to something else 

If there is some other way to go about this or we have gone down the wrong rabbit hole i would appreciate the 2 cents

Thanks!

**Comment by kostis-codefresh** [comment]
> The Question we have is there a way to have this single replica deployment keep in sync with the image being used by the rollout?

The common hack for this is to use Kustomize config generators and make the URL of the backend used by the worker a kustomize parameter. You would need to modify your worker to automatically update its config or using something like Reloader

**Comment by phyzical** [comment]
Thanks for the reply,

Ah okay i think i understand, but to confirm.
we just need to use reloader as the main way that the Deployment will get a new image.

Seems simple enough.  

```mermaid
flowchart LR;
Rollout -->|Changes config map with image ref| ConfigMap
Reloader <-->|Watches for change| ConfigMap
Reloader -->|Updates deployments image spec with image ref|Deployment
```

Thanks!

**Reply by kostis-codefresh** [reply]
@phyzical I published a comprehensive tutorial here https://codefresh.io/blog/multi-service-progressive-delivery-with-argo-rollouts/

**Reply by phyzical** [reply]
Thanks for that 👍 will come in handy
